runtime/near-vm-logic/src/vmstate.rs

```diff
@@ -0,0 +1,177 @@
+use crate::dependencies::MemoryLike;
+use crate::gas_counter::GasCounter;
+
+use near_primitives_core::config::ExtCosts::*;
+use near_primitives_core::config::VMLimitConfig;
+use near_vm_errors::{HostError, VMLogicError};
+
+use core::mem::size_of;
+
+type Result<T> = ::std::result::Result<T, VMLogicError>;
+
+/// Guest memory.
+///
+/// Provides interface to access the guest memory while correctly accounting for
+/// gas usage.
+///
+/// Really the main point of this struct is that it is a separate object so when
+/// its methods are called, such as `memory.get_into(&mut gas_counter, ...)`,
+/// the compiler can deconstruct the access to each field of [`VMLogic`] and do
+/// more granular lifetime analysis.  In particular, this design is what allows
+/// us to forgo copying register value in [`VMLogic::read_register`].
+pub(super) struct Memory<'a>(&'a mut dyn MemoryLike);
+
+macro_rules! memory_get {
+    ($_type:ty, $name:ident) => {
+        pub(super) fn $name(
+            &mut self,
+            gas_counter: &mut GasCounter,
+            offset: u64,
+        ) -> Result<$_type> {
+            let mut array = [0u8; size_of::<$_type>()];
+            self.get_into(gas_counter, offset, &mut array)?;
+            Ok(<$_type>::from_le_bytes(array))
+        }
+    };
+}
+
+macro_rules! memory_set {
+    ($_type:ty, $name:ident) => {
+        pub(super) fn $name(
+            &mut self,
+            gas_counter: &mut GasCounter,
+            offset: u64,
+            value: $_type,
+        ) -> Result<()> {
+            self.set(gas_counter, offset, &value.to_le_bytes())
+        }
+    };
+}
+
+impl<'a> Memory<'a> {
+    pub(super) fn new(mem: &'a mut dyn MemoryLike) -> Self {
+        Self(mem)
+    }
+
+    /// Copies data from guest memory into provided buffer accounting for gas.
+    fn get_into(&self, gas_counter: &mut GasCounter, offset: u64, buf: &mut [u8]) -> Result<()> {
+        gas_counter.pay_base(read_memory_base)?;
+        let len = u64::try_from(buf.len()).map_err(|_| HostError::MemoryAccessViolation)?;
```

Comment:
Note to self: this may seem like a behaviour change, but in practice isn’t, because `usize` is always conversible to `u64` on both 32-bit and 64-bit machines. Rust doesn’t _currently_ have any targets where `usize` is larger than `u64`.